KEYCLOAK-15167 Retrieve email from Twitter IdP

NathanStrobbe opened the pull request:
<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->
I'm updating this issue [KEYCLOAK-15167](https://issues.redhat.com/browse/KEYCLOAK-15167) and opening a new Pull Request closed a week ago here https://github.com/keycloak/keycloak/pull/7510

NathanStrobbe committed:
KEYCLOAK-15167 Retrieve email from Twitter IdP

NathanStrobbe commented on services/src/main/java/org/keycloak/social/twitter/TwitterIdentityProvider.java:
```diff
@@ -190,8 +191,10 @@ public Response authResponse(@QueryParam("state") String state,
+                ConfigurationBuilder builder = new ConfigurationBuilder();
+                builder.setIncludeEmailEnabled(true);
 
+                Twitter twitter = new TwitterFactory().getInstance();
```
Of course! You're right. I added this line in a previous commit but I rebased on an older commit and forgot to add this :sweat_smile:

bjoernhype reviewed (dismissed):
Seems fine for me :), but I didn't test it.
There is one failing JUnit test (JWKTest), but I think thats not related to this MR

bjoernhype commented:
@NathanStrobbe , do you have any idea regarding the failing JUnit test? If that's coming from master, perhaps it is / will be fixed and another rebase could help?

NathanStrobbe commented:
I have no idea... The failure test indicates that line on file `JWKTest.java`
```java
assertEquals(256/8, yBytes.length);
```
I think there is no connection with my code, I'll try to rebase

zak905 commented on services/src/main/java/org/keycloak/social/twitter/TwitterIdentityProvider.java:
```diff
@@ -190,8 +191,10 @@ public Response authResponse(@QueryParam("state") String state,
+                ConfigurationBuilder builder = new ConfigurationBuilder();
+                builder.setIncludeEmailEnabled(true);
 
+                Twitter twitter = new TwitterFactory(builder.build()).getInstance();
```
@NathanStrobbe this could be even shortned to one line: ` Twitter twitter = new TwitterFactory(new ConfigurationBuilder().setIncludeEmailEnabled(true).build()).getInstance();`

NathanStrobbe commented:
Hi @bjoernhype @zak905, I'm updating this PR. Is it possible to approve the workflow run?

bjoernhype commented:
Hi @NathanStrobbe , zak905 already approved. Is there anything I can help to get this merged? Sadly I don't know any "reviewer with write access" :/
It would be great if this simple but relevant issue could be merged

antbig commented:
This is a very useful fix,
Can you have a look @pedroigor ?

NathanStrobbe commented:
Hi @hmlnarik @stianst, could you take a look at this PR to be able to run the workflow please? Thanks.

kaikreuzer commented:
Maybe @pedroigor can help as he just merged [other Twitter related fixes](https://github.com/keycloak/keycloak/pull/8418)? Would be awesome!

Axzial commented:
@mposolda can you review this please ? :)

mposolda commented:
@NathanStrobbe Thanks for the PR! The fix itself looks good to me. Few things:
- Is it possible to rebase this PR and squash the 2 commits into single commit with the JIRA message?
- I will need to make sure that social tests are still working as expected (I have some suspicion that this can be an issue as current test possibly expects that email address is not provided from Twitter). If you have a chance, you can try if twitter related tests in `SocialLoginTest` still work for you on your laptop.

pedroigor commented:
LGTM. If you can address Marek's comments for me we can move forward and have it in.

NathanStrobbe commented:
Hi @pedroigor @mposolda, as suggested I've rebased into 1 commit with the JIRA message. Also, I've launched arquillian test in local and there was a failure because the update profile page didn't show up anymore, cause the email was provided. So I've updated it and it's now successful!

mposolda reviewed (approved):
@NathanStrobbe Thanks!

zak905 commented:
Hello everyone, I have added this additional patch https://github.com/keycloak/keycloak/issues/12912 to avoid having the update profile form when redirecting back from twitter, if the twitter name contains only a single word. `setName` assumes that the name is in form of `[1 to n first names] [last name]`, it is not always the case. Having a null last name causes Keycloak to prompt for it through the update profile form, which is what we wanted to avoid through this PR initially.
